esy-build-package/Boss.ml

```diff
@@ -0,0 +1,162 @@
+module String = struct
+  include Stdlib.String
+  (* The code in this module is copied from JaneStree Base library*)
+  let length = Stdlib.String.length
+  external char_caseless_equal : 'a -> 'a -> bool = "%equal"
+  external unsafe_get : string -> int -> char = "%string_unsafe_get"
+
+  let invalid_argf fmt = Printf.ksprintf (fun s () -> invalid_arg s) fmt
+  let is_substring_at_gen =
+    let rec loop ~str ~str_pos ~sub ~sub_pos ~sub_len ~char_equal =
+      if sub_pos = sub_len
+      then true
+      else if char_equal (unsafe_get str str_pos) (unsafe_get sub sub_pos)
+      then loop ~str ~str_pos:(str_pos + 1) ~sub ~sub_pos:(sub_pos + 1) ~sub_len ~char_equal
+      else false
+    in
+    fun str ~pos:str_pos ~substring:sub ~char_equal ->
+      let str_len = length str in
+      let sub_len = length sub in
+      if str_pos < 0 || str_pos > str_len
+      then
+        invalid_argf
+          "String.is_substring_at: invalid index %d for string of length %d"
+          str_pos
+          str_len
+          ();
+      str_pos + sub_len <= str_len
+      && loop ~str ~str_pos ~sub ~sub_pos:0 ~sub_len ~char_equal
+  let is_prefix_gen string ~prefix ~char_equal =
+    let string_len = length string in
+    let prefix_len = length prefix in
+    string_len >= prefix_len
+    && is_substring_at_gen string ~pos:0 ~substring:prefix ~char_equal
+  let is_prefix s ~prefix = is_prefix_gen s ~prefix ~char_equal:char_caseless_equal
+end
+
+(* The code in the rest of the file is copied from dbuenzli/bos library*)
+
```

Comment:
We need to insert the license header: https://github.com/dbuenzli/bos/blob/master/src/bos.mli#L1

Reply:
This is temporary, I don't plan to have that file merged, this just a POC for discussion